pyscf/lib/linalg_helper.py

```diff
@@ -575,6 +576,77 @@ def davidson1(aop, x0, precond, tol=1e-12, max_cycle=50, max_space=12,
 
     return numpy.asarray(conv), e, x0
 
+def davidson2(aop,x0,precond,nroots=1,tol=1e-12,max_cycle=50,max_space=12,
+              max_memory=MAX_MEMORY,verbose=logger.WARN):
```

Comment:
I can add the docstring to explain that, and I want to explain three points about this. 
1. The function davidson2() serves for a diagonalization solver to solve  a c = e c like davidson1() and davidson_nosym1().  While the sub-space of davidson1() or davidson_nosym1() isn't increased cycles by cycles, the sub-space of davidson2() will be increased until up to the same as the whole a matrix.
2. There is a convergence trouble for spin-flip TDDFT use davidson1() or davidson_nosym1(). While using davidson2(), the roots will be convergent until at a very big sub-space or a same dimension with origin a matrix. 
3. The trouble in point 2 isn't observed in spin-flip TDA.

Reply:
And davidson1()  is used to solve hermitian or real-symmetry matrix by  scipy.linalg.eigh() function, while spin-flip up or spin-flip down TDDFT matrix is not a hermitian or real-symmetry matrix. Davidson_nosym1() serves for GKS or DKS case with complex data. So the davidson2(), as a temporary solver, is added in the file.

Reply:
* What is the difference between davidson2 and davidson1 with a large `max_space`?
* I believe davidson_nosym1 can handle real matrix too. Are there special treatments for the real non-symmetric matrix?

Reply:
1. There is a fake max_space arg in davidson2(), not controlling anything, so the subspace will be increased cycles by cycles. 
2. After seeing the tdscf.ghf and tdscf.gks files, I guess that the structure of init_guess for davidson_nosym1() is like [[x0,y0],[y0,x0.conj()]]. While for spin flip up or down tddft equation, after the two uncoupled, the structure of init_guess is [x0,y0] rather than [[x0,y0],[y0,x0.conj()]], because of the spin flip case.
![sf_init_guess1](https://github.com/pyscf/pyscf/assets/86696701/3b8a835d-16f9-4d7c-ad3e-4b4c31fd4af9)
![sf_init_guess2](https://github.com/pyscf/pyscf/assets/86696701/f5e41bea-3274-4b0b-a420-1dac970bf8c2)